mrp_bom_nested_bom_generator/models/mrp_bom.py

```diff
@@ -0,0 +1,154 @@
+        comodel_name="mrp.bom", inverse_name="parent_bom_id", string="Child BOMs"
+    )
+
+    CHOICE_FUNC = {True: "_create_mrp_bom_record", False: "_append_bom_line_components"}
+
+    @api.depends("nested_bom_ids")
+    def _compute_nested_bom_updated(self):
+        self.write({"nested_bom_updated": True})
+
+    def _prepare_temp_nested_bom_item(self) -> models.Model:
+        """
+        Prepare temp nested bom for current product template
+        :return temp nested bom record
+        :rtype mrp.nested.bom()
+        """
+        return self.nested_bom_ids.new(
+            {
+                "bom_product_tmpl_id": self.product_tmpl_id.id,
+                "product_tmpl_id": self.product_tmpl_id.id,
+                "product_qty": 1,
+                "uom_id": self.product_tmpl_id.uom_id.id,
+            }
+        )
+
+    def group_by_stage(self) -> list:
+        """
+        Group nested bom line by stage
+        :yield list(MRP product, component)
+        :
+        """
+        nestings = [
+            self._prepare_temp_nested_bom_item(),
+        ] + list(self.nested_bom_ids.sorted(lambda line: line.sequence))
+        for i in range(1, len(nestings)):
+            yield nestings[i - 1], nestings[i]
+
+    def unlink_existing_bom(self) -> bool:
+        """
+        Unlink unused BOMs
+        Archive used BOM
+        Unlink all components in current BOM
+        :return: bool
+        """
+        child_bom_ids = self.child_bom_ids
+        if not child_bom_ids:
+            return False
+        mrp_bom_used_ids = (
+            self.env["mrp.production"]
+            .search([("bom_id", "in", child_bom_ids.ids)])
+            .mapped("bom_id")
+        )
+        mrp_bom_unused_ids = child_bom_ids.filtered(
+            lambda bom: bom.id not in mrp_bom_used_ids.ids
+        )
+        mrp_bom_used_ids.sudo().write({"active": False})
+        mrp_bom_unused_ids.sudo().unlink()
+        self.bom_line_ids.sudo().unlink()
+        return True
+
+    def _create_mrp_bom_record(self, nested_bom, lines):
+        """
+        Create mrp.bom record for product
+        :param models.Model nested_bom: mrp.nested.bom record
+        :param list lines: x2m “create” command
+        :return bool: bool
+        """
+        if not lines or not nested_bom:
+            return False
+        if not all(len(line) == 3 for line in lines):
+            return False
+        if not all(not bool(c) and isinstance(v, dict) and v for c, _, v in lines):
+            return False
+        self.env["mrp.bom"].create(
+            {
+                "parent_bom_id": self.id,
+                "product_tmpl_id": nested_bom.product_tmpl_id.id,
+                "product_uom_id": nested_bom.uom_id.id,
+                "product_qty": nested_bom.product_qty,
+                "type": "normal",
+                "bom_line_ids": lines,
+            }
+        )
+        return True
+
+    def _append_bom_line_components(self, lines, **__):
```

Comment:
What is this function used for? And why does it need the “**__”?

Reply:
Resolved